### PR TITLE
Value support for Cassandra types

### PR DESCRIPTION
**Pull request by conorbros**
Pulling some changes out of #404 

### Done

- Support all native and collection Cassandra types in Value
- Add todos for conversions not yet implemented.

### CODE REVIEW

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [inline shotover-proxy/src/message/mod.rs:357]: Do we make another `Value` variant for Decimal or store it in the float one?

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [inline shotover-proxy/src/message/mod.rs:361]: Do we make another Value variant for Decimal or store it in the int one?

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [inline shotover-proxy/src/message/mod.rs:366]: Store Timestamp in the `Bytes` variant?

Comment (conorbros) [inline shotover-proxy/src/message/mod.rs:366]: Date and Time are represented as integers in Cassandra, should they be stored in the Int variant or make a separate one

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [inline shotover-proxy/src/message/mod.rs:286]: Should `Document` and `Map` be merged?

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [reply shotover-proxy/src/message/mod.rs:357]: Decimal can not be losslessly stored as a float.
So I think our only option is to create a Decimal variant.

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [reply shotover-proxy/src/message/mod.rs:366]: If we stored them as ints I dont think we would have a way of telling that they were originally dates.

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:357]: In that case we can use https://docs.rs/bigdecimal/latest/bigdecimal/index.html

**Comment by rukai** [comment]
I think the questions we really need to answer before trying to resolve these issues is:
* Does the data represented in Value need to be able to read without losing precision/accuracy?
* And a stronger requirement: Do we need to be able to roundtrip all data through `Value`?

File: shotover-proxy/src/message/mod.rs
Comment (benbromhead) [reply shotover-proxy/src/message/mod.rs:286]: That would make sense as a Document is a subset of Map

File: shotover-proxy/src/message/mod.rs
Comment (benbromhead) [reply shotover-proxy/src/message/mod.rs:357]: Yup we need a lossless Decimal type (probably a bigint as well)

File: shotover-proxy/src/message/mod.rs
Comment (benbromhead) [reply shotover-proxy/src/message/mod.rs:361]: Let's look at what the Java implementation of Varint is and see if there is a stdlib one in Rust or if we need to implement our own / import a lib

File: shotover-proxy/src/message/mod.rs
Comment (benbromhead) [reply shotover-proxy/src/message/mod.rs:366]: I'm happy for the discriminate on the enum to be the type tag here, but we probably need to make sure we have enough fidelity to represent the most precise version of this... and it's something easily convertible to other time formats

**Comment by benbromhead** [comment]
To answer @rukai questions. Yes to both. 

We should be able to losslesly go from arbitrary Redis <=> Value and Cassandra <=> Value (thus Redis <=> Cassandra)

**Comment by benbromhead** [comment]
This will result in some "less than optimal" decisions. E.g. storing certain types that use more memory than needed. E.g. if one datastores Int type is 32 bits and another is 64 bits.

For the moment, I would just continue down the path of conversion to a common Value type and using that as common intermediary. For some context, however, @Claude-at-Instaclustr and I have also had some discussions about a better way of representing internal types and presenting a nicer way to reason about them. Mainly using a resource descriptor framework (RDF triple). An RDF won't solve the above two questions, but it will provide a better way of reasoning about types... potentially. 

See https://ieeexplore.ieee.org/document/7226706 for some background reading.

**Comment by Claude-at-Instaclustr** [comment]
A better link to the paper: https://arxiv.org/abs/1410.4977

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:322]: Will follow up with this in a later PR when I have a better idea of what parsing and storing a cassandra map into value will look like.

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:404]: We need to be able to differentiate between Counter and Int in order to be lossless.
We can just create a Value::Counter for now and refactor once we add support for a second SQL-like db

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:425]: Collides with ColType::Varchar

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:486]: Lets delete into_bytes instead.
Its unused and I have no idea what its intended use is.
In the current implementation there is no way to revert the data back as there is no tag to identify what the original type was.

**Review by rukai — CHANGES_REQUESTED** [review]
Some initial feedback, I'll look into the BigInt and BigDecimal stuff deeper tomorrow

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:378]: Once we have tests we will find out what parts don't work

Comment (rukai) [reply shotover-proxy/src/message/mod.rs:379]: In that case I guess I wont bother reviewing the conversions themselves I'll just make sure the types we are going between are correct.

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:326]: lets add a `use num::BigInt`

**Review by rukai — CHANGES_REQUESTED** [review]
Once all the feedback is addressed we can land this.

As you've said, in a follow up PR you can add tests for all functions here (Dont bother with into_str_bytes though as its implementation is currently nonsense)

**Comment by conorbros** [comment]
I'll implement the collections stuff in this PR too because that's blocking my other Cassandra integration test PRs

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [inline shotover-proxy/src/message/mod.rs:403]: Shouldn't Ascii have it's own field in Value? Same goes for other Cassandra types that we are currently putting in string. 

@benbromhead

Comment (rukai) [reply shotover-proxy/src/message/mod.rs:403]: yes.
The reason I didnt bring it up in the review was because it was a preexisting issue.
Although thinking about it, it does make sense to resolve it in this PR. (Although still happy for it to go in a follow up)

Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:403]: I'll do it in this PR

**Comment by conorbros** [comment]
Updates: 
 - added support for parsing collections into Value
 - changes to cdrs-tokio: https://github.com/conorbros/cdrs-tokio/commit/0353d89b6b549aa6e5ecdd53f28854e36ab182dd

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:344]: a cassandra bigint appears to just be an i64
https://cassandra.apache.org/doc/latest/cassandra/cql/types.html
lol

Maybe instead of bigint, tinyint etc we should name the integer variants like:
```
I64(i64)
I32(i32)
I16(i16)
I8(i8)
```
Unless the names cassandra uses happen to be standard across many DBs...?

Comment (rukai) [reply shotover-proxy/src/message/mod.rs:344]: moved to the other PR

File: shotover-proxy/Cargo.toml
Comment (rukai) [inline shotover-proxy/Cargo.toml:56]: we'll need to land this cdrs-tokio change before we can merge this PR

File: shotover-proxy/examples/cassandra-passthrough/topology.yaml
Comment (rukai) [inline shotover-proxy/examples/cassandra-passthrough/topology.yaml:5]: why?

Comment (conorbros) [reply shotover-proxy/examples/cassandra-passthrough/topology.yaml:5]: Just a mistake left in while testing

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:457]: What collides?

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:328]: is this different to Varint?

Comment (rukai) [reply shotover-proxy/src/message/mod.rs:328]: ah... maybe this got left in from a bad merge?

Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:328]: Wouldn't we want to know whether it was a BigInt or VarInt when writing back

Comment (rukai) [reply shotover-proxy/src/message/mod.rs:328]: Are you referring to cassandra bigint? I believe we decided to represent that as Integer(_, IntSize::U64).
Cassandra Varint is shotover Varint.

Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:328]: *IntSize::I64, but yes I remember that.

**Comment by conorbros** [comment]
While rebasing I accidentally added `into_str_bytes` back so I have removed it in the most recent commit.

File: shotover-proxy/Cargo.toml
Comment (rukai) [inline shotover-proxy/Cargo.toml:58]: is this an accident to?

Comment (conorbros) [reply shotover-proxy/Cargo.toml:58]: No, I had to put some changes into cdrs-tokio to fix the BigInt issue.

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [reply shotover-proxy/src/message/mod.rs]: They both converted to Value::Strings which would make it impossible to unambiguously return trip.
Problem is resolved now though because ColType::Custom now is now `unimplemented!()`

File: shotover-proxy/Cargo.toml
Comment (rukai) [reply shotover-proxy/Cargo.toml:58]: ah ok, I was confused because it had the same branch name.
Dont forget to make a PR for it.

Comment (conorbros) [reply shotover-proxy/Cargo.toml:58]: Yep, just wanted to confirm it's the right approach before I made a PR.

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:405]: double to string?

File: shotover-proxy/Cargo.toml
Comment (conorbros) [reply shotover-proxy/Cargo.toml:58]: https://github.com/krojew/cdrs-tokio/pull/73

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:618]: This should return Value.
Otherwise its super weird for `Value::create_set(set)` to not return a Value given that create_set belongs to Value

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:607]: This should use into_iter -> map -> collect.
Internally that will call Vec::with_capacity and is just as fast as what you have written here.

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:618]: Lol this seems obvious now that you pointed it out. The caller of this function just wraps it straight away

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:570]: Its unfortunate this duplicates a bunch of logic with build_value_from_cstar_col_type.
Is it not possible to turn the `spec: &ColSpec, data: &CBytes` into a `CassandraType` and call this instead?

File: shotover-proxy/src/message/mod.rs
Comment (conorbros) [inline shotover-proxy/src/message/mod.rs:429]: No longer need this check because cdrs-tokio will make the type CassandraType::Null

**Comment by conorbros** [comment]
Addressed all of the feedback. The changes to `cdrs-tokio` can be found [here](https://github.com/conorbros/cdrs-tokio/tree/typed-collections)

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:471]: This can be written as:
image

Which preallocates the correct capacity.

And same for the other `BtreeMap`/`BTreeSet`s

Comment (conorbros) [reply shotover-proxy/src/message/mod.rs:471]: TIL you can use collect on sets and maps....thought it was just Vec

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:514]: lol, the redundant braces are confusing clippy

File: shotover-proxy/src/message/mod.rs
Comment (rukai) [inline shotover-proxy/src/message/mod.rs:459]: We could drastically simplify this by moving the logic from the functions directly into the match.

image

This removes the need for the panic and the logic is small enough to comfortably fit in the match

**Review by rukai — COMMENTED** [review]
This looks fine to me.
I can approve once the cassandra-protocol changes land upstream.

One thing I did notice: for some reason `#[allow(clippy::mutable_key_type)]` isnt needed if we dont store `values` as an intermediate variable.
Up to you if you want to make use of that.

**Comment by conorbros** [comment]
That is strange, I've removed it so we don't have an unnecessary `#[allow]`.

Opened a PR for the changes to cdrs-tokio: https://github.com/krojew/cdrs-tokio/pull/74

**Review by rukai — APPROVED** [review]
LGTM

**Pull request merged**
